### PR TITLE
Issue-25: Sendgrid Adapter (Lede)

### LINKED ISSUE #25: Sendgrid Adapter (Lede)
### Description

Note: the intent so far is for the Lede development to work on this ticket. For those with access to Athena, here is the ticket: https://getscrum.app/projects/15/stories/13731.

The Lede platform uses Sendgrid as its ESP, and therefore the ESP all publishers on the platform must use as well. We need an adapter for the Newsletter Builder for Sendgrid to make this work seamlessly.

### Use Case

Publishers on Lede will need to be able to do the following, which roughly correspond do the options in the [Single Send](https://docs.sendgrid.com/ui/sending-email/single-sends) functionality in Sendgrid:
* Email subject
* Multi-select of contact lists OR contact segments to which a newsletter will be delivered from among the lists defined in Sendgrid.
* Multi-select of contact lists OR contact segments to exclude from the newsletter delivery (this is probably a nice-to-have).
* Scheduling (when to send the email).
* Select an unsubscribe group for the newsletter. This selection will dictate two things:
  * A list of users that will be excluded from the delivery of this newsletter.
  * The unsubscribe group a user will be added to if they click the "unsubscribe" button in the newsletter footer.

### PR DESCRIPTION
**Pull request by mogmarsh**
### Summary

This pull request introduces the implementation of the Sendgrid Adapter for the Lede platform as described in the issue. It aims to integrate Sendgrid with the Newsletter Builder, allowing publishers on Lede to utilize Sendgrid's functionality seamlessly. This includes handling email subjects, multi-select of contact lists or segments, scheduling, unsubscribe group selection, inline styling for emails, and managing subscription statuses.

Fixes #25

### Changes

- Implemented the Sendgrid adapter to work with the existing Newsletter Builder interface.
- Added functionality for publishers to select multiple contact lists or segments for newsletter delivery.
- Included options for scheduling the delivery time of newsletters.
- Enabled selection of an unsubscribe group, ensuring users can opt-out efficiently and are excluded from future newsletters if they do so.
- Introduced inline styling capabilities for email content to enhance visual presentation.
- Added support for managing subscription statuses, allowing for more precise control over newsletter delivery.
- Implemented suppression list support to manage recipients more effectively, ensuring compliance with unsubscribe requests.

### Test Plan

- Ensure the adapter integrates seamlessly with the current Newsletter Builder setup.
- Verify that all the options provided by the Sendgrid Single Send functionality are accessible and work as expected.
- Test the multi-select feature for contact lists and segments, including the ability to exclude certain lists or segments.
- Check the scheduling functionality to ensure newsletters can be sent at specified times.
- Confirm that selecting an unsubscribe group works as intended, both in excluding users and in managing unsubscribe actions.
- Validate the inline styling functionality for emails to ensure the visual presentation meets expectations.
- Test the management of subscription statuses and suppression lists to confirm efficient handling of recipients and compliance with unsubscribe requests.

### Additional Notes

- Further enhancements to the adapter might be required based on user feedback and additional requirements that may arise during the testing phase.
- It's important to consider any limitations or special configurations required by Sendgrid for the unsubscribe link functionality, as mentioned in the comments.

### CODE REVIEW

File: src/email-providers/class-sendgrid.php [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+
+		try {
+			// @phpstan-ignore-next-line
+			$response = $sg->client->marketing()->singlesends()->_( $campaign_id )->schedule()->put( [ 'send_at' => 'now' ] );
+			return [
+				'response'         => $response->body(),
+				'http_status_code' => $response->statusCode(),
+			];
+		} catch ( \Exception $ex ) {
+			return false;
```
Comment (attackant) [inline src/email-providers/class-sendgrid.php:220]: Do we want any logging or anything for exceptions?

File: src/email-providers/class-sendgrid.php [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+		}
+		$response = $sg->client->marketing()->singlesends()->_( $campaign_id )->get();
+		$body     = (object) json_decode( $response->body() );
+		return [
+			'response'         => [
+				'Status'       => $body->status,
+				'Name'         => $body->name,
+				'Recipients'   => 'N/A',
+				'TotalOpened'  => 'N/A',
+				'UniqueOpened' => 'N/A',
```
Comment (attackant) [inline src/email-providers/class-sendgrid.php:248]: Is N/A a value we ingest elsewhere? Should it be a const?

File: src/email-providers/class-sendgrid.php [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+		return $this->get_senders();
+	}
+
+	/**
+	 * Gets the html content for the newsletter.
+	 *
+	 * @param int $post_id The post id.
+	 * @return string|false
+	 */
+	private function get_content( $post_id ) {
```
Comment (attackant) [inline src/email-providers/class-sendgrid.php:387]: There are places where we don't use return types... is this intentional?

File: src/email-providers/class-sendgrid.php [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+	 * @phpstan-param array{response: mixed, http_status_code: int}|false $result
+	 * @return mixed
+	 */
+	public function get_campaign_id_from_create_result( array|false $result ): mixed {
+		if ( empty( $result ) ) {
+			return false;
+		}
+		$response = $result['response'];
+		if ( ! empty( $response && is_string( $response ) ) ) {
+			$body = (object) json_decode( $response );
```
Comment (attackant) [inline src/email-providers/class-sendgrid.php:284]: Is the cast here needed for a null return?

**Review by attackant — APPROVED** [review]
A few questions, no blockers 🐙

File: src/email-providers/class-sendgrid.php (hunk 4 above)
Comment (mogmarsh) [reply src/email-providers/class-sendgrid.php:284]: the cast was to make phpstan happy

File: src/email-providers/class-sendgrid.php (hunk 2 above)
Comment (mogmarsh) [reply src/email-providers/class-sendgrid.php:248]: ideally we'll get this data from somewhere, but I'm not sure where it lives yet.

File: src/email-providers/class-sendgrid.php (hunk 3 above)
Comment (mogmarsh) [reply src/email-providers/class-sendgrid.php:387]: not intentional

**Pull request merged**